controllers/genericprovider_controller.go

```diff
@@ -76,33 +80,104 @@ func (r *GenericProviderReconciler) Reconcile(ctx context.Context, req reconcile
 
 	defer func() {
 		// Always attempt to patch the object and status after each reconciliation.
-		if err := patchHelper.Patch(ctx, typedProvider.GetObject()); err != nil {
+		if err := patchProvider(ctx, patchHelper, typedProvider); err != nil {
 			reterr = kerrors.NewAggregate([]error{reterr, err})
 		}
 	}()
 
-	// Ignore deleted provider, this can happen when foregroundDeletion
-	// is enabled
-	// Cleanup logic is not needed because owner references set on resource created by
-	// Provider will cause GC to do the cleanup for us.
-	if !typedProvider.GetDeletionTimestamp().IsZero() {
+	// Add finalizer first if not exist to avoid the race condition between init and delete
+	if !controllerutil.ContainsFinalizer(typedProvider.GetObject(), operatorv1.ProviderFinalizer) {
+		controllerutil.AddFinalizer(typedProvider.GetObject(), operatorv1.ProviderFinalizer)
 		return ctrl.Result{}, nil
 	}
 
+	// Handle deletion reconciliation loop.
+	if !typedProvider.GetDeletionTimestamp().IsZero() {
+		return r.reconcileDelete(ctx, typedProvider)
+	}
+
 	return r.reconcile(ctx, typedProvider, typedProviderList)
 }
 
-func (r *GenericProviderReconciler) reconcile(ctx context.Context, genericProvider genericprovider.GenericProvider, genericProviderList genericprovider.GenericProviderList) (_ ctrl.Result, reterr error) {
-	// Run preflight checks to ensure that core provider can be installed properly
-	result, err := preflightChecks(ctx, r.Client, genericProvider, genericProviderList)
-	if err != nil || !result.IsZero() {
-		return result, err
+func patchProvider(ctx context.Context, patchHelper *patch.Helper, provider genericprovider.GenericProvider) error {
+	conds := []clusterv1.ConditionType{
+		operatorv1.PreflightCheckCondition,
+		operatorv1.CertManagerReadyCondition,
+		operatorv1.ProviderInstalledCondition,
+	}
+	conditions.SetSummary(provider, conditions.WithConditions(conds...))
+	return patchHelper.Patch(ctx, provider.GetObject(), patch.WithOwnedConditions{Conditions: append(conds, clusterv1.ReadyCondition)})
+}
+
+func (r *GenericProviderReconciler) reconcile(ctx context.Context, provider genericprovider.GenericProvider, genericProviderList genericprovider.GenericProviderList) (_ ctrl.Result, reterr error) {
```

Comment:
Names returns are generally discouraged, and we don't even seem to be using them, I'd remove them
```suggestion
func (r *GenericProviderReconciler) reconcile(ctx context.Context, provider genericprovider.GenericProvider, genericProviderList genericprovider.GenericProviderList) (ctrl.Result, error) {
```